[WebDriver][BiDi] Implement the browsingContext.getTree method

lauromoura opened the pull request:
#### 09e8854c854800fa8640edd36be6ab1ab1ec0c38
<pre>
[WebDriver][BiDi] Implement the browsingContext.getTree method
<a href="https://bugs.webkit.org/show_bug.cgi?id=283784">https://bugs.webkit.org/show_bug.cgi?id=283784</a>

Reviewed by BJ Burg.

Add support for the root and maxDepth parameter. The root selection is
done by splitting the steps of selecting the navigables to traverse
and actually traversing them.

Also, fix the presence of some fields in the returned payload. In the
current version of the spec, only the &quot;parent&quot; field is actually
optional, while the other fields must be present, even if they&apos;re
null.

In order to support the latter, this commit also adds support for a new
boolean field in object types declared in the automation protocol:
&quot;nullable&quot;. This field generates a &quot;setNull...&quot; method that sets the
related key in the underlying objet with JSON::Value::null(). That way
we can keep the original setter receiving the original types like
Ref&lt;...&gt;.

Originally, the generated inspector code also restricted the
non-optional fields to be set at construction time, with the nested
`Builder` class wrapping the generated type constructor. This commit
also exposes the normal setters for non-optional-but-nullable fields,
so we can set these fields later if needed. For example:

auto obj = Protocol::Module::Object::create()
  .setRequiredField()
  .setNullChildren() // &quot;children&quot; is a &quot;nullable&quot; member
  .release();
...
obj.setChildren(childrenArray);

This &quot;nullable&quot; tag contrasts with &quot;optional&quot; in the sense that with
optional, the key might be missing from the final object.

Note that this doesn&apos;t cover nullable command parameters nor support for
processing of incoming frontend data with nullable fields, as we only
need to support outbound data from the backend (i.e. command return
values) for now.

* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_protocol_types_header.py:
  * Generate new setter models in Builder and actual Model classes.
* Source/JavaScriptCore/inspector/scripts/codegen/models.py:
  * Add &quot;nullable&quot; property to type declarations.
(Protocol.parse_type_member):
(TypeMember.__init__):
* Source/JavaScriptCore/inspector/scripts/tests/expected/type-declaration-object-type.json-result:
  * Rebaseline results.
* Source/JavaScriptCore/inspector/scripts/tests/type-declaration-object-type.json:
  * Add testcase declaration.
* Source/WebKit/UIProcess/Automation/BidiBrowsingContextAgent.cpp:
(WebKit::BidiBrowsingContextAgent::getBrowsingContextID const): Helper
added.
(WebKit::BidiBrowsingContextAgent::getNavigableInfo): Added.
(WebKit::BidiBrowsingContextAgent::getNextTree): Added.
(WebKit::toOptionalUint): Added.
(WebKit::BidiBrowsingContextAgent::getTree): Add support for maxDepth,
root, and actually traverse the navigable trees.
* Source/WebKit/UIProcess/Automation/BidiBrowsingContextAgent.h: Declare
  new methods and types.
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h: Make
  handleForWebFrame* methods public as the BrowsingContext agent uses
  it.
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
  Fix optional and nullable fields for getTree parameters.
* WebDriverTests/TestExpectations.json: Gardening results with this
  patch.

Canonical link: <a href="https://commits.webkit.org/295805@main">https://commits.webkit.org/295805@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/06cbf8faad4ec0a1c3125be95cb9e5fb7f4b9f10

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106185 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25934 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16328 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111382 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56781 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26593 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34437 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80656 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109189 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21019 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95824 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60981 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20572 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13927 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56221 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/98824 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90379 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13962 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114242 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/104802 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33323 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24578 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89731 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33687 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92057 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89430 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22804 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34299 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12113 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28899 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33248 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38660 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/129114 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32994 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35191 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36344 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34592 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_protocol_types_header.py:
```diff
@@ -343,6 +345,14 @@ def _generate_builder_setter_for_member(self, type_member, domain):
         lines.append('        }')
+        if type_member.is_nullable:
+            lines.append('')
+            lines.append('        Builder<STATE | %(camelName)sSet>& setNull%(camelName)s()' % setter_args)
```
Naming alternative: what about `set%(camelName)s*IsNull*()`?

burg commented on Source/WebKit/UIProcess/Automation/BidiBrowsingContextAgent.cpp:
```diff
@@ -125,35 +131,120 @@ void BidiBrowsingContextAgent::create(Inspector::Protocol::BidiBrowsingContext::
+    }
+
+    auto infos = JSON::ArrayOf<Inspector::Protocol::BidiBrowsingContext::Info>::create();
+    getNextTree(WTFMove(navigablePages), WTFMove(infos), WTFMove(maxDepth), [callback = WTFMove(callback)](auto&& result) {
```
This is similar to an async work queue, but the variable names aren't quite aligned with that concept. What about variable names that allow a call like `getNextTree(pagesToProcess, resultsObject, maxDepth, callback)`? (ignoring the necessary moves to illustrate)

burg commented on Source/WebKit/UIProcess/Automation/BidiBrowsingContextAgent.cpp:
```diff
@@ -125,35 +131,120 @@ void BidiBrowsingContextAgent::create(Inspector::Protocol::BidiBrowsingContext::
-    // FIXME: implement `root` option.
-    // FIXME: implement `maxDepth` option.
+    std::optional<uint64_t> maxDepth = std::nullopt;
+    if (optionalMaxDepth) {
```
nice.

burg commented on Source/WebKit/UIProcess/Automation/BidiBrowsingContextAgent.cpp:
```diff
@@ -125,35 +131,120 @@ void BidiBrowsingContextAgent::create(Inspector::Protocol::BidiBrowsingContext::
+        return;
+    }
+
+    Ref webPageProxy = pages.takeLast();
```
Oh interesting, we are traversing most recently appended first? I was expecting FIFO.. 🫥

lauromoura replied:
Oops. Leftover from original code in `WebAutomationSession::getNextContext`

lauromoura committed:
[WebDriver][BiDi] Implement the browsingContext.getTree method
https://bugs.webkit.org/show_bug.cgi?id=283784

Reviewed by BJ Burg.

Add support for the root and maxDepth parameter. The root selection is
done by splitting the steps of selecting the navigables to traverse
and actually traversing them.

Also, fix the presence of some fields in the returned payload. In the
current version of the spec, only the "parent" field is actually
optional, while the other fields must be present, even if they're
null.

In order to support the latter, this commit also adds support for a new
boolean field in object types declared in the automation protocol:
"nullable". This field generates a "setNull..." method that sets the
related key in the underlying objet with JSON::Value::null(). That way
we can keep the original setter receiving the original types like
Ref<...>.

Originally, the generated inspector code also restricted the
non-optional fields to be set at construction time, with the nested
`Builder` class wrapping the generated type constructor. This commit
also exposes the normal setters for non-optional-but-nullable fields,
so we can set these fields later if needed. For example:

auto obj = Protocol::Module::Object::create()
  .setRequiredField()
  .setNullChildren() // "children" is a "nullable" member
  .release();
...
obj.setChildren(childrenArray);

This "nullable" tag contrasts with "optional" in the sense that with
optional, the key might be missing from the final object.

Note that this doesn't cover nullable command parameters nor support for
processing of incoming frontend data with nullable fields, as we only
need to support outbound data from the backend (i.e. command return
values) for now.

* Source/JavaScriptCore/inspector/scripts/codegen/generate_cpp_protocol_types_header.py:
  * Generate new setter models in Builder and actual Model classes.
* Source/JavaScriptCore/inspector/scripts/codegen/models.py:
  * Add "nullable" property to type declarations.
(Protocol.parse_type_member):
(TypeMember.__init__):
* Source/JavaScriptCore/inspector/scripts/tests/expected/type-declaration-object-type.json-result:
  * Rebaseline results.
* Source/JavaScriptCore/inspector/scripts/tests/type-declaration-object-type.json:
  * Add testcase declaration.
* Source/WebKit/UIProcess/Automation/BidiBrowsingContextAgent.cpp:
(WebKit::BidiBrowsingContextAgent::getBrowsingContextID const): Helper
added.
(WebKit::BidiBrowsingContextAgent::getNavigableInfo): Added.
(WebKit::BidiBrowsingContextAgent::getNextTree): Added.
(WebKit::toOptionalUint): Added.
(WebKit::BidiBrowsingContextAgent::getTree): Add support for maxDepth,
root, and actually traverse the navigable trees.
* Source/WebKit/UIProcess/Automation/BidiBrowsingContextAgent.h: Declare
  new methods and types.
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h: Make
  handleForWebFrame* methods public as the BrowsingContext agent uses
  it.
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
  Fix optional and nullable fields for getTree parameters.
* WebDriverTests/TestExpectations.json: Gardening results with this
  patch.

Canonical link: https://commits.webkit.org/295805@main
